framework/validators/DateValidator.php

```diff
@@ -364,35 +373,80 @@ private function parseDateValueFormat($value, $format)
-            $formatter = new IntlDateFormatter($this->locale, IntlDateFormatter::NONE, IntlDateFormatter::NONE, $hasTimeInfo ? $this->timeZone : 'UTC', null, $format);
-        }
+        $formatter = $this->getIntlDateFormatter($format);
         // enable strict parsing to avoid getting invalid date values
         $formatter->setLenient(false);
 
         // There should not be a warning thrown by parse() but this seems to be the case on windows so we suppress it here
         // See https://github.com/yiisoft/yii2/issues/5962 and https://bugs.php.net/bug.php?id=68528
         $parsePos = 0;
         $parsedDate = @$formatter->parse($value, $parsePos);
-        if ($parsedDate === false || $parsePos !== mb_strlen($value, Yii::$app ? Yii::$app->charset : 'UTF-8')) {
+        $valueLength = mb_strlen($value, Yii::$app ? Yii::$app->charset : 'UTF-8');
+        if ($parsedDate === false || $parsePos !== $valueLength || !$this->validateStrictIntl($value, $parsedDate, $formatter)) {
             return false;
         }
 
         return $parsedDate;
     }
 
+    /**
+     * Creates IntlDateFormatter
+     *
+     * @param $format string date format
+     * @return IntlDateFormatter
+     * @throws InvalidConfigException
+     */
+    private function getIntlDateFormatter($format)
+    {
+        if (!isset($this->_dateFormats[$format])) {
+            // if no time was provided in the format string set time to 0 to get a simple date timestamp
+            $hasTimeInfo = (strpbrk($format, 'ahHkKmsSA') !== false);
+            $formatter = new IntlDateFormatter($this->locale, IntlDateFormatter::NONE, IntlDateFormatter::NONE, $hasTimeInfo ? $this->timeZone : 'UTC', null, $format);
+
+            return $formatter;
+        }
+
+        if ($this->type === self::TYPE_DATE) {
+            $dateType = $this->_dateFormats[$format];
+            $timeType = IntlDateFormatter::NONE;
+            $timeZone = 'UTC';
+        } elseif ($this->type === self::TYPE_DATETIME) {
+            $dateType = $this->_dateFormats[$format];
+            $timeType = $this->_dateFormats[$format];
+            $timeZone = $this->timeZone;
+        } elseif ($this->type === self::TYPE_TIME) {
+            $dateType = IntlDateFormatter::NONE;
+            $timeType = $this->_dateFormats[$format];
+            $timeZone = $this->timeZone;
+        } else {
+            throw new InvalidConfigException('Unknown validation type set for DateValidator::$type: ' . $this->type);
+        }
+
+        $formatter = new IntlDateFormatter($this->locale, $dateType, $timeType, $timeZone);
+
+        return $formatter;
+    }
+
+    /**
+     * Performs strict date format validation
+     *
+     * @param $value string date
+     * @param $parsedDate string parsed date
+     * @param IntlDateFormatter $formatter formatter
+     * @return bool
+     */
+    private function validateStrictIntl($value, $parsedDate, IntlDateFormatter $formatter)
+    {
+        if ($this->strictDateFormat) {
```

Comment:
How about moving this condition out of the method?

Reply:
I'm suggesting it because method itself implies that validation is going to be strict.